lib/rack/utf8_sanitizer.rb

```diff
@@ -234,10 +246,15 @@ def unescape_unreserved(input)
     # See also URI::REGEXP::PATTERN::{UNRESERVED,RESERVED}.
     UNSAFE           = /[^\-_.!~*'()a-zA-Z\d;\/?:@&=+$,\[\]%]/
 
-    # Performs the reverse function of `unescape_unreserved`. Unlike
-    # the previous function, we can reuse the logic in URI#encode
+    # Performs the reverse function of `unescape_unreserved`. The logic here is
+    # optimized from URI::RFC2396_Parser#escape
     def escape_unreserved(input)
-      URI::DEFAULT_PARSER.escape(input, UNSAFE)
+      @unsafe_map ||= Hash.new do |table, us|
```

Comment:
This lookup table is for a negation pattern (`UNSAFE = /[^...]/`) so it could match any 1, 2, 3, or 4-byte unicode character. It's possible that this could grow too much as more unique codepoints are encountered.

I'm thinking we could limit the caching to 1 or 2 byte inputs only (up to 65535), e.g.

```
table[us] = encoded if us.length <= 2
```

(Variable names here are from https://ruby-doc.org/stdlib-2.3.1/libdoc/uri/rdoc/URI/RFC2396_Parser.html#method-i-escape although the implementation is a little streamlined.)

Reply:
Went with `bytesize <= 3` here, which seems to be a good tradeoff for size/performance.